File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/commandBased/opmodes/AutoOpMode.java
package org.firstinspires.ftc.teamcode.commandBased.opmodes;

import android.annotation.SuppressLint;

import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName;
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
import org.firstinspires.ftc.robotcore.external.navigation.AxesOrder;
import org.firstinspires.ftc.robotcore.external.navigation.AxesReference;
import org.firstinspires.ftc.robotcore.external.navigation.Orientation;
import org.firstinspires.ftc.teamcode.commandBased.pipelines.AprilTagDetectionPipeline;
import org.firstinspires.ftc.teamcode.commandBased.subsystems.AutoDrivetrainSubsystem;
import org.firstinspires.ftc.teamcode.rr.drive.SampleMecanumDrive;
import org.openftc.apriltag.AprilTagDetection;
import org.openftc.easyopencv.OpenCvCamera;
import org.openftc.easyopencv.OpenCvCameraFactory;
import org.openftc.easyopencv.OpenCvCameraRotation;

import java.util.ArrayList;

public class AutoOpMode extends BaseOpMode {

    protected AutoDrivetrainSubsystem drive;
    protected SampleMecanumDrive rrDrive;

    protected OpenCvCamera camera;
    protected AprilTagDetectionPipeline aprilTagPipeline;
    protected double fx = 578.272;
    protected double fy = 578.272;
    protected double cx = 402.145;
    protected double cy = 221.506;
    protected double tagsize = 0.166;

    protected AprilTagDetection tagOfInterest = null;

    protected double rollingSum = 0;

    protected enum TagPos {
        LEFT(0),
        MIDDLE(1),
        RIGHT(2),
        NULL(3);

        private final double value;

        TagPos(double value) {
            this.value = value;
        }

        public double getValue() {
            return value;
        }
    }

    protected TagPos tagPos = TagPos.NULL;

    @Override
    public void initialize() {
        super.initialize();

        rrDrive = new SampleMecanumDrive(hardwareMap);

        drive = new AutoDrivetrainSubsystem(rrDrive, false);

        aprilTagInitialization();
    }

    @Override
    public void run() {
        super.run();
    }

    protected void aprilTagInitialization() {
        int cameraMonitorViewId = hardwareMap.appContext.getResources().getIdentifier(
                "cameraMonitorViewId",
                "id",
                hardwareMap.appContext.getPackageName()
        );

        camera = OpenCvCameraFactory.getInstance().createWebcam(
                hardwareMap.get(
                        WebcamName.class,
                        "Webcam 1"
                ),
                cameraMonitorViewId
        );

        aprilTagPipeline = new AprilTagDetectionPipeline(tagsize, fx, fy, cx, cy);

        camera.setPipeline(aprilTagPipeline);

        camera.openCameraDeviceAsync(new OpenCvCamera.AsyncCameraOpenListener() {
            @Override
            public void onOpened() {
                camera.startStreaming(800, 448, OpenCvCameraRotation.UPRIGHT);
            }

            @Override
            public void onError(int errorCode) {

            }
        });
    }

    protected void updateCurrentTag() {
        while (!isStarted() && !isStopRequested()) {
            ArrayList<AprilTagDetection> currentDetections = aprilTagPipeline.getLatestDetections();

            if (currentDetections.size() != 0) {
                boolean tagFound = false;
                for (AprilTagDetection tag : currentDetections) {
                    if(tag.id == TagPos.LEFT.value ||
                       tag.id == TagPos.MIDDLE.value ||
                       tag.id == TagPos.RIGHT.value) {
                        tagOfInterest = tag;
                        tagFound = true;
                        break;
                    }
                }
                if (tagFound) {
                    telemetry.addLine("Tag of interest is in sight!\n\nLocation data:");
                    tagToTelemetry(tagOfInterest);
                } else {
                    telemetry.addLine("Don't see tag of interest :(");

                    if (tagOfInterest == null) {
                        telemetry.addLine("(The tag has never been seen)");
                    } else {
                        telemetry.addLine("\nBut we HAVE seen the tag before at:");
                        tagToTelemetry(tagOfInterest);
                    }
                }
            } else {
                telemetry.addLine("Don't see tag of interest");

                if(tagOfInterest == null) {
                    telemetry.addLine("(The tag has never been seen)");
                } else {
                    telemetry.addLine("\nBut we HAVE seen the tag before; last seen at:");
                    tagToTelemetry(tagOfInterest);
                }

            }
            telemetry.update();
        }
        sleep(20);
    }
    
    protected void determinePathFromTag() {
        camera.closeCameraDevice();
        if (tagOfInterest == null) {
            tagPos = TagPos.NULL;
        } else if (tagOfInterest.id == TagPos.LEFT.value) {
            tagPos = TagPos.LEFT;
        } else if (tagOfInterest.id == TagPos.MIDDLE.value) {
            tagPos = TagPos.MIDDLE;
        } else if (tagOfInterest.id == TagPos.RIGHT.value) {
            tagPos = TagPos.RIGHT;
        }
    }

    @SuppressLint("DefaultLocale")
    protected void tagToTelemetry(AprilTagDetection detection)
    {
        Orientation rot = Orientation.getOrientation(detection.pose.R, AxesReference.INTRINSIC, AxesOrder.YXZ, AngleUnit.DEGREES);

        telemetry.addLine(String.format("\nDetected tag ID=%d", detection.id));
        telemetry.addLine(String.format("Translation X: %.2f feet", detection.pose.x));
        telemetry.addLine(String.format("Translation Y: %.2f feet", detection.pose.y));
        telemetry.addLine(String.format("Translation Z: %.2f feet", detection.pose.z));
        telemetry.addLine(String.format("Rotation Yaw: %.2f degrees", rot.firstAngle));
        telemetry.addLine(String.format("Rotation Pitch: %.2f degrees", rot.secondAngle));
        telemetry.addLine(String.format("Rotation Roll: %.2f degrees", rot.thirdAngle));
    }

}

File: TeamCode/src/main/java/teamcode/v1/vision/AutoOpMode.kt
package teamcode.v1.vision

import com.asiankoala.koawalib.command.KOpMode
import com.asiankoala.koawalib.logger.Logger
import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName
import org.firstinspires.ftc.teamcode.koawalib.vision.SleevePipeline
import org.openftc.apriltag.AprilTagDetection
import org.openftc.easyopencv.OpenCvCamera
import org.openftc.easyopencv.OpenCvCamera.AsyncCameraOpenListener
import org.openftc.easyopencv.OpenCvCameraFactory
import org.openftc.easyopencv.OpenCvCameraRotation

open class AutoOpMode : KOpMode(photonEnabled = true) {
    lateinit var camera : OpenCvCamera
    lateinit var pipeline : SleevePipeline

    var fx = 578.272
    var fy = 578.272
    var cx = 402.145
    var cy = 221.506

    var tagsize = 0.166

    var LEFT = 1
    var MIDDLE = 2
    var RIGHT = 3
    var tagOfInterest: AprilTagDetection? = null

    override fun mInit() {
        val cameraMonitorViewId = hardwareMap.appContext.resources.getIdentifier(
            "cameraMonitorViewId",
            "id",
            hardwareMap.appContext.packageName
        )
        camera = OpenCvCameraFactory.getInstance().createWebcam(
            hardwareMap.get(
                WebcamName::class.java, "Webcam"
            ), cameraMonitorViewId
        )
        pipeline = SleevePipeline(tagsize, fx, fy, cx, cy)
        camera.setPipeline(pipeline)
        camera.openCameraDeviceAsync(object : AsyncCameraOpenListener {
            override fun onOpened() {
                camera.startStreaming(800, 448, OpenCvCameraRotation.UPRIGHT)
            }

            override fun onError(errorCode: Int) {}
        })
    }

    override fun mInitLoop() {
        val currentDetections: ArrayList<AprilTagDetection> =
            pipeline.latestDetections
        if (currentDetections.size != 0) {
            var tagFound = false
            for (tag in currentDetections) {
                if (tag.id == LEFT || tag.id == MIDDLE || tag.id == RIGHT) {
                    tagOfInterest = tag
                    tagFound = true
                    break
                }
            }
            if (tagFound) {
                Logger.addTelemetryLine("Tag of interest is in sight!\n\nLocation data:")
                tagToTelemetry(tagOfInterest)
            } else {
                Logger.addTelemetryLine("Don't see tag of interest :(")
                if (tagOfInterest == null) {
                    Logger.addTelemetryLine("(The tag has never been seen)")
                } else {
                    Logger.addTelemetryLine("\nBut we HAVE seen the tag before; last seen at:")
                    tagToTelemetry(tagOfInterest)
                }
            }
        } else {
            Logger.addTelemetryLine("Don't see tag of interest :(")
            if (tagOfInterest == null) {
                Logger.addTelemetryLine("(The tag has never been seen)")
            } else {
                Logger.addTelemetryLine("\nBut we HAVE seen the tag before; last seen at:")
                tagToTelemetry(tagOfInterest)
            }
        }
    }

    override fun mStart() {
        camera.stopStreaming()
    }

    override fun mLoop() {
        if (tagOfInterest != null) {
            Logger.addTelemetryLine("Tag snapshot:\n")
            tagToTelemetry(tagOfInterest)
        } else {
            Logger.addTelemetryLine("No tag snapshot available, it was never sighted during the init loop :(")
        }
    }

    override fun mStop() {
        camera.stopStreaming()
    }

    private fun tagToTelemetry(detection: AprilTagDetection?) {
        Logger.addTelemetryLine(String.format("\nDetected tag ID=%d", detection!!.id))
        Logger.addTelemetryLine(
            String.format(
                "Translation X: %.2f feet",
                detection.pose.x * FEET_PER_METER
            )
        )
        Logger.addTelemetryLine(
            String.format(
                "Translation Y: %.2f feet",
                detection.pose.y * FEET_PER_METER
            )
        )
        Logger.addTelemetryLine(
            String.format(
                "Translation Z: %.2f feet",
                detection.pose.z * FEET_PER_METER
            )
        )
        Logger.addTelemetryLine(
            String.format(
                "Rotation Yaw: %.2f degrees", Math.toDegrees(
                    detection.pose.yaw
                )
            )
        )
        Logger.addTelemetryLine(
            String.format(
                "Rotation Pitch: %.2f degrees", Math.toDegrees(
                    detection.pose.pitch
                )
            )
        )
        Logger.addTelemetryLine(
            String.format(
                "Rotation Roll: %.2f degrees", Math.toDegrees(
                    detection.pose.roll
                )
            )
        )
    }

    companion object {
        const val FEET_PER_METER = 3.28084
    }
}

